### DISCUSSION TITLE
CUDA (nvcc) cross compiler

### DISCUSSION THREAD

**Discussion by dat-tran-matternet** [OE4T/meta-tegra#947] [Q&A] [answered]
I'm trying to get a cross compiler for CUDA nvcc ( cross compile cuda code for jetson on my x86 host). I tried the following in my image but don't see the cross compiiler for nvcc. thanks

```
TOOLCHAIN_TARGET_TASK_append_tegra = "\
    cuda-compiler \
    cuda-libraries \
    cuda-nvcc \
    cuda-nvcc-dev \
    opencv-dev \
 "
```

I build the Target SDK like so:  `bitbake -c populate_sdk $IMAGE`

**Comment by ichergui** [comment, marked as answer]
Hey 

Use `TOOLCHAIN_HOST_TASK` instead of `TOOLCHAIN_TARGET_TASK`

* Add the following line in your `conf/local.conf` file

```
TOOLCHAIN_HOST_TASK_append = " nativesdk-packagegroup-cuda-sdk-host"
```

**Reply by dat-tran-matternet** [reply]
Got it. thanks for your help!

**Reply by dat-tran-matternet** [reply]
One more question:  I'm trying to cross compile for the Jetson. Do I have to do anything extra to tell nvcc to forward C/C++ code to cross compiler aarch64-poky-linux-gcc/g++ ?  I don't want it to use the native gcc/g++.  If it uses ${CC} ${CXX} variables then it should do what I want but not sure how nvcc works. thanks

Update: seems like nvcc has options to set the compiler. I'll try that. thanks

**Reply by ichergui** [reply]
`NVCC` needs special options for the compilation.
`NVCC` should be used only when you have cuda code.

Take a look in the [nvidia cuda comipler documentation](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#purpose-of-nvcc) to get more insights

**Reply by madisongh** [reply]
If you use cmake's CUDA language support for your project, the SDK environment should take care of this for you. Otherwise, take a look at the environment variables we set as part of the SDK setup, once you get the host-side packagegroup into your SDK -- I believe the `CUDAFLAGS` variable (which you can reference in a makefile) woul dhave the necessary settings.

**Reply by dat-tran-matternet** [reply]
Thanks for your help guys. I was able to get it working with running:  
`nvcc main.cu $CUDAFLAGS -ccbin=${TARGET_PREFIX}g++ -O3 -o cuda-test`

I plan to use CMake in the future but want to make sure my setup is correct with a simple hello world example. 
Thanks again.

**Reply by madisongh** [reply]
Ah yeah, you do need to pass `-ccbin` as well.  The environment variable `CUDAHOSTCXX` can be used as the value for that.

**Reply by dat-tran-matternet** [reply]
You guys are awesome!  Here is the updated command with your suggestions:  

`nvcc main.cu $CUDAFLAGS -ccbin=${CUDAHOSTCXX} -O3 -o cuda-test`

**Reply by ichergui** [reply]
I'm happy that all works fine for you.
